Installing libraries

In [1]:
import pandas as pd
from scipy.io import arff

T2. Load the dataset in Python using the ARFF importer from the scipy library.


In [2]:
# Function to load an ARFF
def read_arff_as_df_scipy():
    dt = arff.loadarff("/content/dataset_55_hepatitis.arff")
    df = pd.DataFrame(dt[0], columns=dt[1])
    return df

**T3.** *Have a look at the dataset by means of methods info() and head(): find the column which stores class data (ground truth), determine the number of samples and features of this dataset, detect columns with missing*


In [3]:
df = read_arff_as_df_scipy()
df.head()

AGE        SEX STEROID ANTIVIRALS FATIGUE MALAISE ANOREXIA LIVER_BIG  \
0  30.0    b'male'   b'no'      b'no'   b'no'   b'no'    b'no'     b'no'   
1  50.0  b'female'   b'no'      b'no'  b'yes'   b'no'    b'no'     b'no'   
2  78.0  b'female'  b'yes'      b'no'  b'yes'   b'no'    b'no'    b'yes'   
3  31.0  b'female'    b'?'     b'yes'   b'no'   b'no'    b'no'    b'yes'   
4  34.0  b'female'  b'yes'      b'no'   b'no'   b'no'    b'no'    b'yes'   

  LIVER_FIRM SPLEEN_PALPABLE SPIDERS ASCITES VARICES  BILIRUBIN  \
0      b'no'           b'no'   b'no'   b'no'   b'no'        1.0   
1      b'no'           b'no'   b'no'   b'no'   b'no'        0.9   
2      b'no'           b'no'   b'no'   b'no'   b'no'        0.7   
3      b'no'           b'no'   b'no'   b'no'   b'no'        0.7   
4      b'no'           b'no'   b'no'   b'no'   b'no'        1.0   

   ALK_PHOSPHATE   SGOT  ALBUMIN  PROTIME HISTOLOGY    Class  
0           85.0   18.0      4.0      NaN     b'no'  b'LIVE'  
1          135.0   42.0      3.5      NaN     b'no'  b'LIVE'  
2           96.0   32.0      4.0      NaN     b'no'  b'LIVE'  
3           46.0   52.0      4.0     80.0     b'no'  b'LIVE'  
4            NaN  200.0      4.0      NaN     b'no'  b'LIVE'

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   AGE              155 non-null    float64
 1   SEX              155 non-null    object 
 2   STEROID          155 non-null    object 
 3   ANTIVIRALS       155 non-null    object 
 4   FATIGUE          155 non-null    object 
 5   MALAISE          155 non-null    object 
 6   ANOREXIA         155 non-null    object 
 7   LIVER_BIG        155 non-null    object 
 8   LIVER_FIRM       155 non-null    object 
 9   SPLEEN_PALPABLE  155 non-null    object 
 10  SPIDERS          155 non-null    object 
 11  ASCITES          155 non-null    object 
 12  VARICES          155 non-null    object 
 13  BILIRUBIN        149 non-null    float64
 14  ALK_PHOSPHATE    126 non-null    float64
 15  SGOT             151 non-null    float64
 16  ALBUMIN          139 non-null    float64
 17  PROTIME         

Dataset info:
* The class data (groun truth) is  into the column '*Class*'
* Number of samples: 155
* Number of features: 19 (20 including the Class column)
* Columns with missing:
  * PROTIME: Has 88 samples due to NaN values.
  * STEROID: Has different fields with a ?.
  * ALK_PHOSPHATE: Has NaN values.


In [5]:
 # MORE INFO ABOUT MISSING DATA
print("Null data:")
print(df.isnull().sum())
print("------")
print("Na data:")
print(df.isna().sum())

Null data:
AGE                 0
SEX                 0
STEROID             0
ANTIVIRALS          0
FATIGUE             0
MALAISE             0
ANOREXIA            0
LIVER_BIG           0
LIVER_FIRM          0
SPLEEN_PALPABLE     0
SPIDERS             0
ASCITES             0
VARICES             0
BILIRUBIN           6
ALK_PHOSPHATE      29
SGOT                4
ALBUMIN            16
PROTIME            67
HISTOLOGY           0
Class               0
dtype: int64
------
Na data:
AGE                 0
SEX                 0
STEROID             0
ANTIVIRALS          0
FATIGUE             0
MALAISE             0
ANOREXIA            0
LIVER_BIG           0
LIVER_FIRM          0
SPLEEN_PALPABLE     0
SPIDERS             0
ASCITES             0
VARICES             0
BILIRUBIN           6
ALK_PHOSPHATE      29
SGOT                4
ALBUMIN            16
PROTIME            67
HISTOLOGY           0
Class               0
dtype: int64


**T4.** *Adopt the following cleaning and filling strategy for dealing with missing values:*
* *Drop features with more than 40% of missing values.*
* *As strategy #1, drop samples with missing values and store the resulting dataframe as df1. Check the
number of available samples and features.*
* *As strategy #2, fill features with missing values using the median for real-valued data and the mode for
categorical data and store the resulting dataframe as df2. Check the number of available samples and
features.*

El código threshold = len(df) * (1 - threshold) es una forma de calcular el número mínimo de valores no nulos que una columna debe tener para no ser eliminada. Aquí está lo que hace cada parte:

len(df): Esto devuelve el número total de filas en el DataFrame df. Es el número total de muestras que tienes.

1 - threshold: Dado que threshold se especifica como la fracción de valores que pueden estar faltantes (por ejemplo, 0.4 para un 40%), 1 - threshold sería la fracción de valores que deben estar presentes. Por ejemplo, si aceptas un 40% de valores faltantes, entonces deseas que el 60% de los valores estén presentes.

Al multiplicar estos dos valores, obtienes el número mínimo absoluto de valores no nulos que deben estar presentes en cada columna para que esa columna no sea eliminada.

Por ejemplo, si tienes 100 filas de datos y has decidido que puedes tolerar un máximo del 40% de valores faltantes en cualquier columna, entonces:

threshold es 0.4.
1 - threshold es 0.6, lo que significa que esperas que al menos el 60% de los datos estén presentes.
len(df) * (1 - threshold) será 100 * 0.6 = 60, indicando que cada columna debe tener al menos 60 valores no nulos para no ser eliminada.
Este cálculo se utiliza en el método dropna de Pandas para especificar el número mínimo de entradas no nulas requeridas a lo largo de un eje (en este caso, cada columna, que es el eje 1) para evitar que la columna sea eliminada.



In [8]:
treshold = 0.4  # There's a typo here; it should be "threshold".
total_filas = len(df)  # Number of samples in the dataframe.
valores_presentes = 1 - treshold  # Calculates the proportion of present values required.
min_absolute_values_no_null = total_filas * valores_presentes  # Calculates the absolute number of non-null values required.

# Drops columns with less non-null values than the threshold.
df_dropna_cleaned = df.dropna(thresh=min_absolute_values_no_null, axis=1)

df1 = df_dropna_cleaned.dropna()  # Drops rows with any missing values.

# Creates a copy of the cleaned dataframe for the second strategy.
df2 = df_dropna_cleaned.copy()

# Select columns with real numerical data.
real_data = df2.select_dtypes(include=['float64', 'int']).columns

# Select columns with categorical data.
categorical_data = df2.select_dtypes(include=['object', 'bool']).columns

# Fill missing values for real numerical data with the median of the column.
for column in real_data:
    df2[column].fillna(df2[column].median(), inplace=True)

# Fill missing values for categorical data with the mode of the column.
for column in categorical_data:
    df2[column].fillna(df2[column].mode()[0], inplace=True)

print("Dataframe after deleting features with more than 40% of missing values")
# Corrected from `df_dropna_info` to `df_dropna_cleaned`
print(f"Remaining features: {df_dropna_cleaned.shape[1]}")
# Corrected from `df_dropna_info` to `df_dropna_cleaned`
print(f"Available samples: {df_dropna_cleaned.shape[0]}")

print("\nStrategy 1: Deleting samples with missing values:")
# Swapped the shape indices because shape[1] is the number of columns (features),
# and shape[0] is the number of rows (samples).
print(f"Remaining features in df1: {df1.shape[1]}")
print(f"Remaining samples in df1: {df1.shape[0]}")

print("\nStrategy 2: Filling features with missing values:")
# Again, swapped the shape indices.
print(f"Remaining features in df2: {df2.shape[1]}")
print(f"Remaining samples in df2: {df2.shape[0]}")


Dataframe after deleting features with more than 40% of missing values
Remaining features: 19
Available samples: 155

Strategy 1: Deleting samples with missing values:
Remaining features in df1: 19
Remaining samples in df1: 120

Strategy 2: Filling features with missing values:
Remaining features in df2: 19
Remaining samples in df2: 155


**T5.** *Transform categorical features into integer-valued features, using progressive integer labels. You have to do it for both df1 and df2.*

Categorical features often need to be transformed into a numeric format to be used in machine learning models. This task uses the LabelEncoder from sklearn to transform categorical columns into integer labels.

In [10]:
from sklearn.preprocessing import LabelEncoder  # Import the LabelEncoder class for encoding labels.

def encode_categorical(df):
    le = LabelEncoder()                         # Instantiate the label encoder.
    for col in df.columns:                      # Iterate over each column in the DataFrame.
        if df[col].dtype == 'object':           # Check if the column is of object type (categorical).
            df[col] = le.fit_transform(df[col]) # Encode the categorical values to integers.
    return df                                   # Return the DataFrame with encoded columns.

df1_encoded = encode_categorical(df1.copy())    # Apply encoding to df1 and create df1_encoded.
df2_encoded = encode_categorical(df2.copy())    # Apply encoding to df2 and create df2_encoded.
print(df1_encoded.head())
print(df2_encoded.head())

    AGE  SEX  STEROID  ANTIVIRALS  FATIGUE  MALAISE  ANOREXIA  LIVER_BIG  \
0  30.0    1        1           0        0        0         0          1   
1  50.0    0        1           0        1        0         0          1   
2  78.0    0        2           0        1        0         0          2   
3  31.0    0        0           1        0        0         0          2   
5  34.0    0        2           0        0        0         0          2   

   LIVER_FIRM  SPLEEN_PALPABLE  SPIDERS  ASCITES  VARICES  BILIRUBIN  \
0           1                1        1        1        1        1.0   
1           1                1        1        1        1        0.9   
2           1                1        1        1        1        0.7   
3           1                1        1        1        1        0.7   
5           1                1        1        1        1        0.9   

   ALK_PHOSPHATE  SGOT  ALBUMIN  HISTOLOGY  Class  
0           85.0  18.0      4.0          0      1  
1     

**T6.** *Normalize features using min-max normalization for df1 and df2*.

Feature normalization is a scaling technique that brings all values into the range [0, 1]. This is important for models that are sensitive to large numerical values. This task uses the MinMaxScaler from sklearn to normalize all features.


In [11]:
from sklearn.preprocessing import MinMaxScaler  # Import the MinMaxScaler class for normalization.

def normalize_features(df):
    scaler = MinMaxScaler()                     # Instantiate the MinMaxScaler.
    df_scaled = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)  # Scale the features and create a new DataFrame.
    return df_scaled                            # Return the normalized DataFrame.

df1_normalized = normalize_features(df1_encoded)  # Normalize df1_encoded to create df1_normalized.
df2_normalized = normalize_features(df2_encoded)  # Normalize df2_encoded to create df2_normalized.
print(df1_normalized.head())
print(df2_normalized.head())


        AGE  SEX  STEROID  ANTIVIRALS  FATIGUE  MALAISE  ANOREXIA  LIVER_BIG  \
0  0.323944  1.0      0.5         0.0      0.0      0.0       0.0        0.5   
1  0.605634  0.0      0.5         0.0      1.0      0.0       0.0        0.5   
2  1.000000  0.0      1.0         0.0      1.0      0.0       0.0        1.0   
3  0.338028  0.0      0.0         1.0      0.0      0.0       0.0        1.0   
4  0.380282  0.0      1.0         0.0      0.0      0.0       0.0        1.0   

   LIVER_FIRM  SPLEEN_PALPABLE  SPIDERS  ASCITES  VARICES  BILIRUBIN  \
0         0.5              0.5      0.5      0.5      0.5   0.155556   
1         0.5              0.5      0.5      0.5      0.5   0.133333   
2         0.5              0.5      0.5      0.5      0.5   0.088889   
3         0.5              0.5      0.5      0.5      0.5   0.088889   
4         0.5              0.5      0.5      0.5      0.5   0.133333   

   ALK_PHOSPHATE      SGOT  ALBUMIN  HISTOLOGY  Class  
0       0.219331  0.007782  0.

**T7.** *Feature selection using forward and backward sequential selection.*

SequentialFeatureSelector is used with an SVM classifier using a linear kernel to identify which features contribute most to predicting the target variable. This is done both in a forward and backward manner, where forward selection starts with no features and adds one at a time, and backward selection starts with all features and removes one at a time.

In [13]:
from sklearn.feature_selection import SequentialFeatureSelector  # Import SequentialFeatureSelector for feature selection.
from sklearn.svm import SVC                                     # Import SVC class for creating the selection model.

def feature_selection(df, y_column_name):
    X = df.drop(columns=[y_column_name])                         # Separate the feature matrix from the labels.
    y = df[y_column_name]                                        # Isolate the label vector.

    sfs_forward = SequentialFeatureSelector(SVC(kernel='linear'),  # Create a forward sequential feature selector.
                                            n_features_to_select=5,
                                            direction='forward').fit(X, y)
    sfs_backward = SequentialFeatureSelector(SVC(kernel='linear'),  # Create a backward sequential feature selector.
                                             n_features_to_select=5,
                                             direction='backward').fit(X,y)   # Fit the backward feature selector to the features and labels.

    df_forward = X[sfs_forward.get_feature_names_out()]  # Select the features chosen by the forward method.
    df_backward = X[sfs_backward.get_feature_names_out()]  # Select the features chosen by the backward method.

    return df_forward, df_backward  # Return the DataFrames with selected features.

# Replace 'target' with your actual target column name
actual_target_column_name = 'Class'  # Replace 'Class' with your actual target column name.

df3, df4 = feature_selection(df1_normalized, actual_target_column_name)  # Apply feature selection to df1_normalized.
df5, df6 = feature_selection(df2_normalized, actual_target_column_name)  # Apply feature selection to df2_normalized.

print("Features selected by forward selection from df1_normalized:", df3.columns.tolist())
print("Features selected by backward selection from df1_normalized:", df4.columns.tolist())
print("Features selected by forward selection from df2_normalized:", df5.columns.tolist())
print("Features selected by backward selection from df2_normalized:", df6.columns.tolist())



Features selected by forward selection from df1_normalized: ['AGE', 'MALAISE', 'LIVER_BIG', 'ASCITES', 'BILIRUBIN']
Features selected by backward selection from df1_normalized: ['MALAISE', 'ASCITES', 'ALK_PHOSPHATE', 'ALBUMIN', 'HISTOLOGY']
Features selected by forward selection from df2_normalized: ['LIVER_FIRM', 'ASCITES', 'VARICES', 'BILIRUBIN', 'ALBUMIN']
Features selected by backward selection from df2_normalized: ['MALAISE', 'LIVER_FIRM', 'ASCITES', 'ALBUMIN', 'HISTOLOGY']


**T8.** *Dimensionality reduction using PCA*.

PCA, or Principal Component Analysis, reduces the dimensionality of data by transforming it into a new set of variables, the principal components, which are uncorrelated and which capture the maximum amount of variance.


In [14]:
from sklearn.decomposition import PCA  # Import PCA for dimensionality reduction.

def reduce_dimensions(df, n_components=5):
    pca = PCA(n_components=n_components)  # Instantiate the PCA model with the number of desired components.
    principal_components = pca.fit_transform(df)  # Fit PCA on the dataset and apply the dimensionality reduction.
    df_pca = pd.DataFrame(data=principal_components)  # Create a new DataFrame with the principal components.
    return df_pca  # Return the new DataFrame with reduced dimensions.

df1_pca = reduce_dimensions(df1_normalized, 5)  # Apply PCA to reduce to 5 components.
df2_pca = reduce_dimensions(df2_normalized, 5)

print(df1_pca.head())  # Display the first few rows of the DataFrame with PCA applied to df1_normalized.
print(df2_pca.head())  # Display the first few rows of the DataFrame with PCA applied to df2_normalized.


          0         1         2         3         4
0 -0.868923  0.100085  0.224141 -0.011926  0.287704
1 -0.291080  0.179293  0.207746 -0.403591 -0.266730
2 -0.375175  0.162059  0.126583 -0.570216 -0.272636
3 -0.965299  0.214908 -0.488233  0.650456 -0.267424
4 -0.905275 -0.104778  0.311070 -0.030111  0.002677
          0         1         2         3         4
0 -0.638351  0.079683  0.088872 -0.050766  0.599008
1 -0.407662  0.102966  0.312726 -0.048114  0.226962
2 -0.459878  0.064887  0.138737 -0.283198 -0.082283
3 -0.750604 -0.266761  0.018817  0.746232 -0.348014
4 -0.580482  0.250066  0.086069 -0.300892 -0.225183


**T9.** *Model performance comparison*

 Support Vector Machine (SVM) model with a Radial Basis Function (RBF) kernel is trained and evaluated. The train_test_split function from sklearn.model_selection is used to divide the dataset into a training set and a test set. The SVM is then trained on the training set and evaluated on the test set. The accuracy score is calculated by comparing the predicted labels against the true labels in the test set.


In [15]:
# Import necessary libraries
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Function to train and evaluate the SVM model
def train_and_evaluate_models(df, y_column_name):
    # Separate the features from the target variable
    X = df.drop(columns=[y_column_name])
    y = df[y_column_name]

    # Split the dataset into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Instantiate the SVM model with specific hyperparameters
    svm = SVC(C=1e6, kernel='rbf', gamma='auto')

    # Fit the SVM model to the training data
    svm.fit(X_train, y_train)

    # Predict the target variable for the test set
    y_pred = svm.predict(X_test)

    # Calculate and return the accuracy of the model
    return accuracy_score(y_test, y_pred)

# Example of using the train_and_evaluate_models function
# Replace 'Class' with the actual name of your target column
accuracy_df1 = train_and_evaluate_models(df1_normalized, 'Class')
accuracy_df2 = train_and_evaluate_models(df2_normalized, 'Class')

# Print the accuracy scores
print("Accuracy of model with df1_normalized:", accuracy_df1)
print("Accuracy of model with df2_normalized:", accuracy_df2)

# Assuming you have df3, df4, df5, df6 after feature selection from Task 7,
# you would similarly call the function for those dataframes and print their accuracies.



Accuracy of model with df1_normalized: 0.875
Accuracy of model with df2_normalized: 0.7419354838709677
